gh-132042: Prebuild mro_dict for find_name_in_mro to speedup class creation

sergey-miryanov opened the pull request:
This is one of the optimizations from https://github.com/python/cpython/pull/132156 that moved to separate PR.

All three optimizations from original PR give about 40% speedup on tests. 
This optimization give about 15%-18% speedup.

<!-- gh-issue-number: gh-132042 -->
* Issue: gh-132042
<!-- /gh-issue-number -->

sergey-miryanov committed:
Prebuild mro_dict for find_name_in_mro

sergey-miryanov committed:
Simplify find_name_in_mro_new and add some comments

sergey-miryanov committed:
Fix comment for find_namd_in_mro becase it returns strong ref now

vstinner commented on Objects/typeobject.c:
```diff
@@ -11314,11 +11335,42 @@ fixup_slot_dispatchers(PyTypeObject *type)
+    // Try to prebuild MRO dict. If we fails then clear mro_dict and
+    // reset error flag because we don't expect any exceptions. If
+    // fails to prebuild MRO dict then update_on_slot will use
+    // previous version of find_name_in_mro.
```
Please modify fixup_slot_dispatchers() to report the error to its caller. The only caller is type_new_impl() which can already report errors (return NULL).

sergey-miryanov replied:
Ok. Will do.

sergey-miryanov replied:
Fixed.

sergey-miryanov replied:
Augh, I forgot that `test.test_descr.MiscTests.test_type_lookup_mro_reference` fails if we use prebuilt mro-dict. I'm trying to fix.

sergey-miryanov replied:
I just reverted commit that allows fixup_slot_dispatchers to fail, because I don't see a good way to handle exceptions that can be raised from iterating base class.

vstinner replied:
> I don't see a good way to handle exceptions that can be raised from iterating base class

What do you mean? There is no need to "handle exceptions", just report them to the caller, no? What's the problem?

vstinner replied:
> Augh, I forgot that test.test_descr.MiscTests.test_type_lookup_mro_reference fails if we use prebuilt mro-dict. I'm trying to fix.

You can update the test using:

```py
        class MyKey(object):
            def __hash__(self):
                return hash('mykey')

            def __eq__(self, other):
                try:
                    X.__bases__ = (Base2,)
                except NameError:
                    pass
```

sergey-miryanov replied:
> What do you mean? There is no need to "handle exceptions", just report them to the caller, no? What's the problem?

According to your suggestion to fix the test there are no problems - I was confused trying to distinguish between errors from dict's merge and errors from tests and decided that it can't be done reliably. If we fix the test, then all is ok. Thanks!

sergey-miryanov replied:
@vstinner Fixed.

vstinner commented on Objects/typeobject.c:
```diff
@@ -11314,11 +11335,42 @@ fixup_slot_dispatchers(PyTypeObject *type)
+                break;
+            }
+        } else {
+            if (PyDict_Merge(mro_dict, dict, 1) < 0) {
```
I'm not sure that the dict is merged in the right direction. It should be `PyDict_Merge(dict, mro_dict, 1)` to keep find_name_in_mro() ordrer, no?

sergey-miryanov replied:
PyDict_Merge merges b(dict) to a(mro_dict). 

Also, we merge from bottom base to the top (n-i-1 above). In my benchmarks it was faster if we copy bottom base and then merge to the top, instead of copy top base and then merge rest. I think it is because bottom base is an object that have a lot of entries in dict.

vstinner replied:
Oh ok, I see. The dict is merged in the right direction (type inheritance). Thanks for the explanation.

vstinner reviewed (commented):
> All optimizations give about 40% speedup on tests.

Would you mind to run benchmarks on this PR?

First, rebase the PR on the main branch to retrieve the "Do not lookup tp_dict each time to speedup class creation" change.

sergey-miryanov committed:
Merge branch 'main' into gh-132042-precalc-mro-dict

sergey-miryanov committed:
Allow fixup_slot_dispatchers fails

sergey-miryanov committed:
Merge branch 'main' into gh-132042-precalc-mro-dict

sergey-miryanov committed:
Revert "Allow fixup_slot_dispatchers fails"

This reverts commit a64d3a0ee5011fbf13301d829b811513b7e03638.

sergey-miryanov committed:
Update comment about prebuild MRO-dict

sergey-miryanov committed:
Merge branch 'main' into gh-132042-precalc-mro-dict

sergey-miryanov committed:
Allow fixup_slot_dispatchers fails

sergey-miryanov committed:
Fix test_type_lookup_mro_reference test because it raises exception while type initialized

- in the old realisation this exception swallowed and this base not checked
  while finding in mro. so we don't change observed behavior with this
  change.

sergey-miryanov committed:
Merge branch 'main' into gh-132042-precalc-mro-dict

vstinner commented on Objects/typeobject.c:
```diff
@@ -11315,21 +11338,60 @@ update_slot(PyTypeObject *type, PyObject *name)
 
+    PyObject *mro = Py_NewRef(lookup_tp_mro(type));
+
+    // Try to prebuild MRO dict. We build it in bottom-top manner,
```
"Try to prebuild MRO dict." it's not an attempt, if an error occurs, everything is aborted.

```suggestion
    // Build MRO dict. We build it in bottom-top manner,
```

sergey-miryanov replied:
Thanks!

vstinner reviewed (approved):
LGTM. I just have a remark on a comment.

I'm surprised that the latest benchmark run no longer shows slower tests.

sergey-miryanov committed:
Update comment about building MRO dict

Co-authored-by: Victor Stinner <vstinner@python.org>

sergey-miryanov committed:
Merge branch 'main' into gh-132042-precalc-mro-dict

sergey-miryanov committed:
Merge branch 'gh-132042-precalc-mro-dict' of github.com:sergey-miryanov/cpython into gh-132042-precalc-mro-dict

sergey-miryanov committed:
Remove END_TYPE_LOCK from fixup_slot_dispatchers

sergey-miryanov committed:
Merge branch 'main' into gh-132042-precalc-mro-dict

sergey-miryanov committed:
Trying to fix build

sergey-miryanov committed:
Fix build

sergey-miryanov committed:
Merge branch 'main' into gh-132042-precalc-mro-dict

sergey-miryanov committed:
Fix merge conflicts

sergey-miryanov committed:
Fix merge conflicts

vstinner commented on Objects/typeobject.c:
```diff
@@ -11506,7 +11503,7 @@ fixup_slot_dispatchers(PyTypeObject *type)
     Py_DECREF(mro);
 
-    END_TYPE_LOCK();
+    // END_TYPE_LOCK();
```
If the code is no longer needeed, please remove it instead of putting it in a comment.

sergey-miryanov replied:
Yes, I just forgot to delete it. Thanks! 

Btw, it is an outdated code - I revert back because https://github.com/python/cpython/pull/131174 was reverted.

sergey-miryanov committed:
Merge branch 'main' into gh-132042-precalc-mro-dict

sergey-miryanov committed:
Replace lock with assert

sergey-miryanov committed:
Fix fixup_slot_dispatchers definition

sergey-miryanov committed:
Revert "Replace lock with assert"

This reverts commit b3874ce7b09d3b284881d2800291f87d50f7d487.

sergey-miryanov committed:
Replace lock with assert
